aesara/tensor/math_opt.py

```diff
@@ -3523,6 +3532,36 @@ def local_reciprocal_1_plus_exp(fgraph, node):
                         return out
 
 
+def distribute_mul_over_add(in_lv, out_lv):
+    from kanren import conso, eq, fact, heado, tailo
+
+    # This does the optimization A * (x + y + z) = A * x + A * y + A * z
+    A_lv, add_term_lv, add_cdr_lv, dot_cdr_lv, add_flat_lv = lvars(5)
+    fact(commutative, at.mul)
+    fact(associative, at.mul)
+    fact(commutative, at.add)
+    fact(associative, at.add)
+    return lall(
+        # Make sure the input is a `at.mul`
+        eq(in_lv, etuple(at.mul, A_lv, add_term_lv)),
+        # Make sure the term being `at.mul`ed is an `add`
+        heado(at.add, add_term_lv),
+        # Flatten the associative pairings of `add` operations
+        assoc_flatten(add_term_lv, add_flat_lv),
+        # Get the flattened `add` arguments
+        tailo(add_cdr_lv, add_flat_lv),
+        # Add all the `at.mul`ed arguments and set the output
+        conso(at.add, dot_cdr_lv, out_lv),
+        # Apply the `at.mul` to all the flattened `add` arguments
+        mapo(lambda x, y: conso(at.mul, etuple(A_lv, x), y), add_cdr_lv, dot_cdr_lv),
+    )
+
+
+distribute_mul_over_add_opt = KanrenRelationSub(
+    lambda x, y: distribute_mul_over_add(y, x)
```

Comment:
As of now, I could not figure out how to do the reverse using `kanren` but I guess this works just as well.

Reply:
Are you referring to the argument swap?

Reply:
No, I mean I'm yet to figure out how to do xa + ya + z*a -> (x + y + z)*a purely in kanren (without the argument swap). (Though that might not be necessary.)

Reply:
If the relation encoded by the `kanren` goal `distribute_mul_over_add` is `x*a + y*a + z*a = (x + y + z)*a`, in that exact left-hand/right-hand side order, then `distribute_mul_over_add(l, r)` will unify `l` with `x*a + y*a + z*a` and/or `r` with `(x + y + z)*a`: i.e. the first argument, `l`, will yield "expanded" results, and the second argument, `r`, will yield the "collected" results.